### PR TITLE
Add tests, change main.py, change serializer.py

### PR DESCRIPTION
**Pull request by anastasiakoroto**
#### Tests
- added tests
- functions in **serializer_test.py** and **deserializer_test.py** changed  (`@pytest.mark.parametrize()` used to avoid code repeat)
- mock tests corrected
- tests on getting valid/invalid json data from web/file added
#### Serializer
- changed response to exceptions in `is_url()` function (**main.py**)
- added *commas in commas* and *backslash* cases to new function `replace_symbol()` (**serializer.py**)
- added verify existence of **`'\n'`** of input string (**serializer.py**)
- view of `is_url(), url_argument(), file_argument()` functions changed a bit (helps to avoid unexpected errors, displays information about the estimated nature of errors)

### CODE REVIEW

File: JSON_serializer/Tests/check_valid_test.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            main.requests.get(mock_requests_get)
+
+    @mock.patch.object(main.requests, "get", side_effect=TooManyRedirects)
+    def test_is_url_redirects_except(self, mock_requests_get=
+                                   'https://www.amazon.in/Thor-Hammer-Logo-Keychain-Metal'
+                                   '/dp/B01F1QGBHI/ref=sr_1_3?keywords=thor&qid=1564739314&s=electronics&sr=1-3'):
+        with pytest.raises(TooManyRedirects):
+            main.requests.get(mock_requests_get)
+
+    @mock.patch.object(main.requests, "get", side_effect=RequestException)
```
Comment (melnichevv) [inline JSON_serializer/Tests/check_valid_test.py]: these tests look good, but you should add more tests: like getting json from web, getting incorrect json, etc.
```Python
@patch("json_serializer.src.serializer.helpers.requests")
def test_valid_web_response(mock_requests) -> None:
    mock_response = """{
            "fruit": "Apple",
            "size": "Large",
            "color": "Red"
    }"""
    mock_requests.get.return_value = Mock(text=mock_response)
    mock_requests.head.return_value = Mock(headers={"content-type": "application/json"})

    url = "https://support.oneskyapp.com/hc/en-us/article_attachments/202761627/example_1.json"
    # Here you should call your app with this url and check that result is expected
```

File: JSON_serializer/Tests/check_valid_test.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    def test_is_url_err(self):
+        inp = main.is_url('https://www.harrypotter.com')
+        out = False
+        assert inp == out
+
+    @mock.patch.object(main.requests, "get", side_effect=Timeout)
+    def test_is_url_timeout(self, mock_requests_get='https://peggycarter.org'):
+        with pytest.raises(Timeout):
+            main.requests.get(mock_requests_get)
```
Comment (melnichevv) [inline JSON_serializer/Tests/check_valid_test.py]: you shouldn't call `main.requests.get` here. You should execute your app here instead.
Something like this:

- Mock main.requests.get
- Call your app (inside it `requests.get` should be used
- Check the result based on the mocking

**Review by melnichevv — CHANGES_REQUESTED** [review]

File: JSON_serializer/Tests/check_valid_test.py (hunk 1 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: OK. Will add it.

File: JSON_serializer/Tests/check_valid_test.py (hunk 2 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: Will rewrite it.

File: JSON_serializer/Tests/check_valid_test.py (hunk 1 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: Corrected (attempt no.2)

File: JSON_serializer/Tests/check_valid_test.py (hunk 2 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: Corrected (attempt no.2)

File: JSON_serializer/Tests/check_valid_test.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+from unittest import mock
+from requests import ConnectionError, Timeout, RequestException, TooManyRedirects
+
+import main
+
+
+class TestValid:
+
+    def test_is_valid_file(self):
+        inp = main.is_valid_file('/home/anastasiya/Downloads/ironman.txt')
```
Comment (melnichevv) [inline JSON_serializer/Tests/check_valid_test.py]: I'm pretty sure that this test will fail on other machine. You should use mock here

File: JSON_serializer/Tests/check_valid_test.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+    def test_json_from_url(self):
+        with mock.patch('main.requests') as mock_requests:
+            mock_requests.get.return_value.text = """  {
+                                                         "fruit": "Apple",
+                                                         "size": "Large",
+                                                         "color": "Red"
+                                                       }
+            """
+            url_address = 'https://support.oneskyapp.com/hc/en-us/article_attachments/202761627/example_1.json'
```
Comment (melnichevv) [inline JSON_serializer/Tests/check_valid_test.py]: there is no need in using such URLs. You can simply use any dummy URL like "https://dummy.com"

File: JSON_serializer/Tests/deserializer_test.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                                 ('{"database": "HowlingCommandos\\\\Survivor\\\\Bucky \"Winter Soldier\" Barnes"}',
+                                  {'database': 'HowlingCommandos\\Survivor\\Bucky \"Winter Soldier\" Barnes'})
+                             ])
+    def test_deserialize_symbols(self, t_inp, j_out):
+        des_inp = obj.deserialize(t_inp)
+        json_out = json.loads(json.dumps(j_out))
+        assert des_inp == json_out
+
+
+if __name__ == '__main__':
```
Comment (melnichevv) [inline JSON_serializer/Tests/deserializer_test.py]: You don't need this. You should be able to run your tests by simple `pytest Tests`. If that's not - you should fix this

**Review by melnichevv — COMMENTED** [review]
Overall looks good. Left a few comments.
Add `requirements.txt` file please and make sure that it contains only required packages

File: JSON_serializer/Tests/check_valid_test.py (hunk 3 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: Agree. I will correct it.

File: JSON_serializer/Tests/deserializer_test.py (hunk 5 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/deserializer_test.py]: It works with `pytest Tests`. I just thought it should be in program. OK, will remove it.

File: JSON_serializer/Tests/check_valid_test.py (hunk 3 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: Corrected (attempt no.3)

File: JSON_serializer/Tests/check_valid_test.py (hunk 4 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/check_valid_test.py]: Corrected (attempt no.3)

File: JSON_serializer/Tests/deserializer_test.py (hunk 5 above)
Comment (anastasiakoroto) [reply JSON_serializer/Tests/deserializer_test.py]: Corrected (attempt no.3)

**Review by melnichevv — APPROVED** [review]
LGTM!

**Pull request merged**